### ISSUE TITLE
`dag.Host().Directory(".")` doesn't return root directory defined in `dagger.json`

### ISSUE THREAD

**Issue by aweris** [dagger/dagger#5937]
### What is the issue?

When configuring `dagger.json` with a parent directory that uses `".."` to access repository code using `dag.Host().Directory(".")`, it still returns the directory where my module code exists.

```json
{
  "name": "hello",
  "sdk": "go",
  "root": "..",
  "sdkRuntime": "vito/dagger-sdk-go:no-ptrs@sha256:2bb26bae3f30b920a4530b7495127fe4986102b06a99ff170500e83c25739b27"
}
```

module code:

```go
type Hello struct{}

func (m *Hello) World(ctx context.Context) error {
	dir := dag.Host().Directory(".")

	entries, err := dir.Entries(ctx)
	if err != nil {
		return err
	}

	for _, entry := range entries {
		fmt.Println(entry)
	}

	return nil
}
```

Directory tree:
```shell
➜  dummy tree
.
└── hello
    ├── dagger.gen.go
    ├── dagger.json
    ├── go.mod
    ├── go.sum
    ├── main.go
    └── querybuilder
        ├── marshal.go
        └── querybuilder.go

3 directories, 7 files
```

Run Command : 

```shell
echo "{hello{world}}"|dagger query
```

List of entries:

```shell
25: exec /runtime
25: [0.19s] dagger.gen.go
25: [0.19s] dagger.json
25: [0.19s] go.mod
25: [0.19s] go.sum
25: [0.19s] main.go
25: [0.19s] querybuilder
25: exec /runtime DONE
```

### Log output

```shell
➜  hello echo "{hello{world}}"|dagger query --progress=plain
1: connect
1: > in init
1: starting engine 
1: starting engine [0.09s]
Connected to engine github.com/dagger/dagger (version devel ())
1: starting session 
1: [0.12s] OK!
1: starting session [0.02s]
1: connect DONE

3: dagger query --progress=plain
3: loading module 
3: ...

7: upload /Users/aweris/development/workspaces/github.com/zenith/dummy from aweris-2.local (client id: pcivjgxz87n81kpld251099mn) DONE
7: > in host.directory /Users/aweris/development/workspaces/github.com/zenith/dummy
7: upload /Users/aweris/development/workspaces/github.com/zenith/dummy from aweris-2.local (client id: pcivjgxz87n81kpld251099mn) DONE

7: upload /Users/aweris/development/workspaces/github.com/zenith/dummy from aweris-2.local (client id: pcivjgxz87n81kpld251099mn)
7: > in host.directory /Users/aweris/development/workspaces/github.com/zenith/dummy
7: transferring /Users/aweris/development/workspaces/github.com/zenith/dummy: 
7: transferring /Users/aweris/development/workspaces/github.com/zenith/dummy: 124KiB [0.04s]
7: upload /Users/aweris/development/workspaces/github.com/zenith/dummy from aweris-2.local (client id: pcivjgxz87n81kpld251099mn) DONE

6: upload /Users/aweris/development/workspaces/github.com/zenith/dummy from aweris-2.local (client id: pcivjgxz87n81kpld251099mn) CACHED
6: > in host.directory /Users/aweris/development/workspaces/github.com/zenith/dummy
6: upload /Users/aweris/development/workspaces/github.com/zenith/dummy from aweris-2.local (client id: pcivjgxz87n81kpld251099mn) CACHED

8: blob://sha256:c1b70c391fbd0fa438b44c699d6937a445962ffdf2709d1323605b309af8f233 DONE
8: > in host.directory /Users/aweris/development/workspaces/github.com/zenith/dummy
8: blob://sha256:c1b70c391fbd0fa438b44c699d6937a445962ffdf2709d1323605b309af8f233 DONE

8: blob://sha256:c1b70c391fbd0fa438b44c699d6937a445962ffdf2709d1323605b309af8f233 CACHED
8: > in host.directory /Users/aweris/development/workspaces/github.com/zenith/dummy
8: blob://sha256:c1b70c391fbd0fa438b44c699d6937a445962ffdf2709d1323605b309af8f233 CACHED

3: dagger query --progress=plain
3: ...

11: upload /usr/local/share/dagger from 4425d3f6d6df (client id: idmtqjgbazpihe9y1v79a6vbl) (include: go-module-sdk-image.tar) DONE
11: > in load builtin module sdk go
11: upload /usr/local/share/dagger from 4425d3f6d6df (client id: idmtqjgbazpihe9y1v79a6vbl) (include: go-module-sdk-image.tar) DONE

11: upload /usr/local/share/dagger from 4425d3f6d6df (client id: idmtqjgbazpihe9y1v79a6vbl) (include: go-module-sdk-image.tar)
11: > in load builtin module sdk go
11: transferring /usr/local/share/dagger: 
11: transferring /usr/local/share/dagger: 47B [0.00s]
11: upload /usr/local/share/dagger from 4425d3f6d6df (client id: idmtqjgbazpihe9y1v79a6vbl) (include: go-module-sdk-image.tar) DONE

10: upload /usr/local/share/dagger from 4425d3f6d6df (client id: idmtqjgbazpihe9y1v79a6vbl) (include: go-module-sdk-image.tar) CACHED
10: > in load builtin module sdk go
10: upload /usr/local/share/dagger from 4425d3f6d6df (client id: idmtqjgbazpihe9y1v79a6vbl) (include: go-module-sdk-image.tar) CACHED

12: blob://sha256:0a9d9cf2cfb793633215bc1d399bef514dcf89b1f7f1f7ce3824e174bcd70033 DONE
12: > in load builtin module sdk go
12: blob://sha256:0a9d9cf2cfb793633215bc1d399bef514dcf89b1f7f1f7ce3824e174bcd70033 DONE

12: blob://sha256:0a9d9cf2cfb793633215bc1d399bef514dcf89b1f7f1f7ce3824e174bcd70033 CACHED
12: > in load builtin module sdk go
12: blob://sha256:0a9d9cf2cfb793633215bc1d399bef514dcf89b1f7f1f7ce3824e174bcd70033 CACHED

13: oci-layout://dagger/import@sha256:a6e07eb7243021fec5ad7072f368d4da8f1305a493b20649f44ee8835523bf82
13: > in load builtin module sdk go
13: resolve dagger/import@sha256:a6e07eb7243021fec5ad7072f368d4da8f1305a493b20649f44ee8835523bf82 
13: resolve dagger/import@sha256:a6e07eb7243021fec5ad7072f368d4da8f1305a493b20649f44ee8835523bf82 [0.00s]
13: oci-layout://dagger/import@sha256:a6e07eb7243021fec5ad7072f368d4da8f1305a493b20649f44ee8835523bf82 DONE

13: oci-layout://dagger/import@sha256:a6e07eb7243021fec5ad7072f368d4da8f1305a493b20649f44ee8835523bf82 CACHED
13: > in load builtin module sdk go
13: oci-layout://dagger/import@sha256:a6e07eb7243021fec5ad7072f368d4da8f1305a493b20649f44ee8835523bf82 CACHED

15: mkdir / DONE
15: > in load builtin module sdk go
15: mkdir / DONE

14: mkfile /input.json CACHED
14: > in load builtin module sdk go
14: mkfile /input.json CACHED

16: exec /runtime CACHED
16: > in load builtin module sdk go
16: exec /runtime CACHED

17: exec go build -o /runtime . CACHED
17: > in load builtin module sdk go
17: exec go build -o /runtime . CACHED

18: exec /usr/local/bin/codegen --module . --propagate-logs=true CACHED
18: > in load builtin module sdk go
18: exec /usr/local/bin/codegen --module . --propagate-logs=true CACHED

3: dagger query --progress=plain
3: loading module [0.35s]
3: ...

21: query
21: ...

24: mkfile /input.json CACHED
24: mkfile /input.json CACHED

17: exec go build -o /runtime . CACHED
17: > in load builtin module sdk go
17: exec go build -o /runtime . CACHED

17: exec go build -o /runtime . CACHED
17: > in load builtin module sdk go
17: exec go build -o /runtime . CACHED

26: exec go build -o /runtime . CACHED
26: exec go build -o /runtime . CACHED

27: exec /usr/local/bin/codegen --module . --propagate-logs=true CACHED
27: exec /usr/local/bin/codegen --module . --propagate-logs=true CACHED

25: exec /runtime
25: ...

1: connect DONE
1: > in init
1: > in init
1: starting engine [0.00s]
1: starting session [0.10s]
1: [0.10s] OK!
1: connect DONE

32: upload . from buildkitsandbox (client id: 8q2bwrntmt5oa8u1f6kya5ors) DONE
32: > in host.directory .
32: upload . from buildkitsandbox (client id: 8q2bwrntmt5oa8u1f6kya5ors) DONE

32: upload . from buildkitsandbox (client id: 8q2bwrntmt5oa8u1f6kya5ors)
32: > in host.directory .
32: transferring .: 
32: transferring .: 119.4KiB [0.00s]
32: upload . from buildkitsandbox (client id: 8q2bwrntmt5oa8u1f6kya5ors) DONE

31: upload . from buildkitsandbox (client id: 8q2bwrntmt5oa8u1f6kya5ors) CACHED
31: > in host.directory .
31: upload . from buildkitsandbox (client id: 8q2bwrntmt5oa8u1f6kya5ors) CACHED

33: blob://sha256:ec233242ea4d805997b2e58c5aadc7cea763c0991cf90a9ee9f78b306a495ab5 DONE
33: > in host.directory .
33: blob://sha256:ec233242ea4d805997b2e58c5aadc7cea763c0991cf90a9ee9f78b306a495ab5 DONE

33: blob://sha256:ec233242ea4d805997b2e58c5aadc7cea763c0991cf90a9ee9f78b306a495ab5 CACHED
33: > in host.directory .
33: blob://sha256:ec233242ea4d805997b2e58c5aadc7cea763c0991cf90a9ee9f78b306a495ab5 CACHED

25: exec /runtime
25: [0.19s] dagger.gen.go
25: [0.19s] dagger.json
25: [0.19s] go.mod
25: [0.19s] go.sum
25: [0.19s] main.go
25: [0.19s] querybuilder
25: exec /runtime DONE

3: dagger query --progress=plain
3: ...

21: query
21: [0.22s] {
21: [0.22s]     "hello": {
21: [0.22s]         "world": null
21: [0.22s]     }
21: [0.22s] }
21: query DONE

3: dagger query --progress=plain
3: dagger query --progress=plain DONE
➜  hello 
```

### SDK version

vito/dagger-sdk-go:no-ptrs@sha256:2bb26bae3f30b920a4530b7495127fe4986102b06a99ff170500e83c25739b27

### OS version

macOS 14.0

**Comment by aweris** [comment]
I guess this is related as well 

```
➜  gale git:(main) ✗ dagger mod use ./../docker
✘ asModule(sourceSubpath: ".") ERROR [0.00s]
• Engine: d86d308ec69e (version v0.9.0)
⧗ 1.13s ✔ 7 ∅ 1 ✘ 1
Error: failed to automate vcs: failed to get vcs ignored paths: input:1: host.directory.asModule failed to create module from config `::.`: failed to get dependency mod from ref "../docker": failed to get config file: lstat /docker/dagger.json: no such file or directory
```

**Comment by sipsma** [comment]
@aweris we made a change to the working dir of modules in v0.9.0.

It used to be the if you have a separate `root` value, then that would be your current working directory, but now your current working directory is always where the module's `dagger.json` is.

However, you still have full access to the root dir, it's just above you.

E.g. if:
1. your module is located at `/foo/bar/blah/dagger.json`
2. you set root to be `../../` (i.e. `/foo`)
Then in the module container, your working dir will be `/src/bar/blah`, with the full contents of `/foo` at `/src`.

We made this change because it feels much simpler for anyone new coming into Zenith for the first time (though I totally get it's probably confusing if you've already gotten used to the previous behavior). Basically, you can always assume your current working directory will be the same one that contains your modules dagger.json; the workdir isn't going to change just because you changed root. Changing root only affects what parent directories are available to you, which reflects the reason we needed to add support for root in the first place.

---

There is one annoyance with all this currently,  tracked in this issue: https://github.com/dagger/dagger/issues/5914

Basically, you often want to call something like `dag.Host().Directory("..")` in this sort of setup, but our API currently has a restriction on making calls like that left over from a long time ago. We'll fix that, but in the mean time you can use a workaround like this: https://github.com/sipsma/dagger/blob/c847458d6df1e184487e3a29f6d00a86ef67e661/sdk/python/runtime/main.go#L81C1-L88

**Comment by aweris** [comment]
Perfect, thank you for sharing the workaround.  As long as I have way to access parent, I don't need to use root at all 🙏

**Comment by TomChv** [comment]
@aweris Should I close this issue now since you're problem is fixed?

**Comment by aweris** [comment]
I can workaround the initial issue as @sipsma described but `dagger mod use ./../docker` is still not working. But I can create a separate issue for that if you want.

**Comment by TomChv** [comment]
> @TomChv I can workaround the initial issue as @sipsma described but dagger mod use ./../docker is still not working.

Strange, yeah it should be something different, you can check an example of local import on this repository: https://github.com/quartz-technology/daggerverse

The `ci` module import all others and it works

**Comment by aweris** [comment]
Yeps you're right it's something different, Clean up all cache, images, and generating modules from scratch worked. Probably, it related some misconfiguration I missed somewhere. Since it's all clear. Closing the issue
